### DISCUSSION TITLE
openSubsonicExtensions shape is not representable in XML

### DISCUSSION THREAD

**Discussion by sentriz** [opensubsonic/open-subsonic-api#56] [General]
hey, have we thought about the openSubsonicExtensions response shape in XML? to me it looks like it's not possible to 1-1 translate it. (related, the shape in Go is map[string][]int, which the Go xml marshaler will error on)

**Comment by Tolriq** [comment]
I did not, but since the first one is not yet validated we can still rework that. I guess extension being an object too? 
```json
[ { "name": "transcodeOffset", "versions" : [ 1 ] }, { "name": "template", "versions" : [ 1, 3 ] }]
```

**Comment by sentriz** [comment]
yes, that looks like it works for both :+1: 

---snip---

**Reply by spl0k** [reply]
I may be mistaken, but IIRC multiple attributes with the same name aren't allowed in XML, so your example isn't valid.

**Reply by sentriz** [reply]
it is allowed, and it happens a lot, for example in search3 we have 

---snip---

**Reply by Tolriq** [reply]
According to XML standard no it seems it's not allowed to have multiple attribute name with the same name. The versions should be a list inside the openSubsonicExtensions  element.

**Reply by sentriz** [reply]
@Tolriq did you see my comment above? it is valid XML, and we do exactly this for the search3 endpoint for example

**Reply by spl0k** [reply]
Your last comment show tags, which is indeed allowed. I'm talking about attributes which are the `name="value"` within tags. Your first comment shows multiple `versions` attributes.

**Reply by sentriz** [reply]
oh the attributes! sorry my mistake

in that case, adjusting the type

```json
{
 "subsonic-response": {
  "status": "ok",
  "version": "1.15.0",
  "type": "gonic",
  "serverVersion": "0.15.2",
  "openSubsonic": true,
  "openSubsonicExtensions": [
   { "name": "someExample", "versions": [ 1, 2, 3 ] },
   { "name": "transcodeOffset", "versions": [ 1 ] }
  ]
 }
}
```

```xml
<subsonic-response status="ok" version="1.15.0" xmlns="http://subsonic.org/restapi" type="gonic" serverVersion="0.15.2" openSubsonic="true">
 <openSubsonicExtensions name="someExample">
  <versions>1</versions>
  <versions>2</versions>
  <versions>3</versions>
 </openSubsonicExtensions>
 <openSubsonicExtensions name="transcodeOffset">
  <versions>1</versions>
 </openSubsonicExtensions>
</subsonic-response>
```

```go
type OpenSubsonicExtension struct {
	Name     string `xml:"name,attr" json:"name"`
	Versions []int  `xml:"versions"  json:"versions"`
}

type OpenSubsonicExtensions []OpenSubsonicExtension
```

**Reply by Tolriq** [reply]
I've updated the PR.
